### ISSUE TITLE
[bug] music widget crash

### ISSUE THREAD

**Issue by J4gQBqqR** [MM2-0/Kvaesitso#9]
Currently, the music widget does not allow picking specific music player.
In the situation where the default music player is unclear, it will crash upon opening. Can the default music player be specified by user? 

```
android.app.PendingIntent$CanceledException
	at android.app.PendingIntent.send(PendingIntent.java:886)
	at android.app.PendingIntent.send(PendingIntent.java:698)
	at de.mm20.launcher2.ui.widget.MusicWidgetKt$MusicWidget$1$3.invoke(MusicWidget.kt:125)
	at de.mm20.launcher2.ui.widget.MusicWidgetKt$MusicWidget$1$3.invoke(MusicWidget.kt:121)
	at androidx.compose.foundation.ClickableKt$combinedClickable$4$gesture$1$4.invoke-k-4lQ0M(Clickable.kt:339)
	at androidx.compose.foundation.ClickableKt$combinedClickable$4$gesture$1$4.invoke(Clickable.kt:318)
	at androidx.compose.foundation.gestures.TapGestureDetectorKt$detectTapGestures$2$1$1.invokeSuspend(TapGestureDetector.kt:125)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:178)
	at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:166)
	at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:397)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:431)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$default(CancellableContinuationImpl.kt:420)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:328)
	at androidx.compose.ui.input.pointer.SuspendingPointerInputFilter$PointerEventHandlerCoroutine.offerPointerEvent(SuspendingPointerInputFilter.kt:511)
	at androidx.compose.ui.input.pointer.SuspendingPointerInputFilter.dispatchPointerEvent(SuspendingPointerInputFilter.kt:406)
	at androidx.compose.ui.input.pointer.SuspendingPointerInputFilter.onPointerEvent-H0pRuoY(SuspendingPointerInputFilter.kt:419)
	at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:310)
	at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:297)
	at androidx.compose.ui.input.pointer.NodeParent.dispatchMainEventPass(HitPathTracker.kt:179)
	at androidx.compose.ui.input.pointer.HitPathTracker.dispatchChanges(HitPathTracker.kt:98)
	at androidx.compose.ui.input.pointer.PointerInputEventProcessor.process-BIzXfog(PointerInputEventProcessor.kt:80)
	at androidx.compose.ui.platform.AndroidComposeView.sendMotionEvent-8iAsVTc(AndroidComposeView.android.kt:1159)
	at androidx.compose.ui.platform.AndroidComposeView.handleMotionEvent-8iAsVTc(AndroidComposeView.android.kt:1109)
	at androidx.compose.ui.platform.AndroidComposeView.dispatchTouchEvent(AndroidComposeView.android.kt:1059)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2799)
	at com.android.internal.policy.DecorView.superDispatchTouchEvent(DecorView.java:488)
	at com.android.internal.policy.PhoneWindow.superDispatchTouchEvent(PhoneWindow.java:1873)
	at android.app.Activity.dispatchTouchEvent(Activity.java:4125)
	at androidx.appcompat.view.WindowCallbackWrapper.dispatchTouchEvent(WindowCallbackWrapper.java:69)
	at com.android.internal.policy.DecorView.dispatchTouchEvent(DecorView.java:446)
	at android.view.View.dispatchPointerEvent(View.java:14568)
	at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:6022)
	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:5825)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5316)
	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5373)
	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5339)
	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:5491)
	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5347)
	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:5548)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5320)
	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:5373)
	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:5339)
	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:5347)
	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:5320)
	at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:8086)
	at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:8037)
	at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:7998)
	at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:8209)
	at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:220)
	at android.os.MessageQueue.nativePollOnce(Native Method)
	at android.os.MessageQueue.next(MessageQueue.java:335)
	at android.os.Looper.loop(Looper.java:183)
	at android.app.ActivityThread.main(ActivityThread.java:7664)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

**Comment by MM2-0** [comment]
The player will try to launch the activity associated with the current media session, or, if no media session is active, the last playing music app (by package name). I think the crash you mentioned above will be fixed in the next update (though I can't reproduce it - what exactly did you do?).

You can also choose which player to open by long pressing the album cover.

**Comment by J4gQBqqR** [comment]
I have multiple music players and the widget always open a specific one. 
I uninstalled that music player it always opens and the widget crashed.

**Comment by MM2-0** [comment]
It should open the player you last used. Unless the player is not a music player. Only apps that declare the "android.intent.category.APP_MUSIC" category in their AndroidManifest are considered as music players.
Does it fix the problem if you play music in any one of the other apps?

**Comment by J4gQBqqR** [comment]
That could be the problem.
As you can see, I played with Finmap and it still crashed. Could it be because of Finmap is not announcing itself as APP_MUSIC?

![Screenshot_20211222-083918_Kvæsitso](https://user-images.githubusercontent.com/77554817/147101634-cac07a7b-56d3-4347-9f6a-4b7f8e0b8682.png)

**Comment by MM2-0** [comment]
Seems to be the problem. I checked the Finamp source and it does not declare "android.intent.category.APP_MUSIC". So probably I should find a better way to filter for music apps. The problem is, I don't want browser apps and YouTube to appear on my homescreen. Maybe a manual selection of the preferred music app would indeed be the best solution.

**Comment by MM2-0** [comment]
I think this is resolved in 1.4
